### ISSUE TITLE
SBML models with `NaN` volumes result in `initialAmount=NaN` and integration error

### ISSUE THREAD

**Issue by matthiaskoenig** [sys-bio/roadrunner#923]
Hi all,
here some strange bug when working with models with NaN volumes which are only set before integration. 
When having a model with compartment `size=NaN` on model loading an `initialAmount=NaN` is generated for the currentSBML even if there are already valid `initialConcentration` set for the species. This results in subsequent integration errors due to NaNs.

`original SBML`:
```
<species metaid="meta_dex_ext" sboTerm="SBO:0000247" id="dex_ext" name="dextromethorphan (plasma)" 
compartment="Vext" initialConcentration="0" substanceUnits="mmole" hasOnlySubstanceUnits="false" boundaryCondition="false" constant="false">
```

`getCurrentSBML`:
```
 <species metaid="meta_dex_ext" sboTerm="SBO:0000247" id="dex_ext" name="dextromethorphan (plasma)" 
compartment="Vext" initialAmount="NaN" substanceUnits="mmole" hasOnlySubstanceUnits="false" boundaryCondition="false" constant="false">
```

Example attached:
```python
import roadrunner

r: roadrunner.RoadRunner = roadrunner.RoadRunner("nan_species.xml")

# setting a valid volume, now the model is completely specified and can be
# numerically integrated
r.setValue("Vext", 1.0)

# Instead of `initialConcentration="0.0" defined in the model there is an
# incorrect `initialAmount="NaN"` on the `dex_ext` species probably due to some 
# pre-processing steps of the model
sbml_str = r.getCurrentSBML()
print("-" * 80)
print(sbml_str)
print("-" * 80)

# As a consequence the model does not integrate
s = r.simulate(start=0, end=10, steps=10)
```
Resulting in
```
Error: CVODE Error: CV_CONV_FAILURE, Module: CVODES, Function: CVode, Message: At t = 0 and h = 1.3093e-10, the corrector convergence test failed repeatedly or with |h| = hmin.
```

Here the files
[nan_species.zip](https://github.com/sys-bio/roadrunner/files/7629742/nan_species.zip)

**Comment by luciansmith** [comment]
It seems like everything is working to me.  If a species only has an initial concentration, the amount is determined by multiplying the concentration by the compartment size.  Roadrunner only works with amounts, so it gets NaN, just as it was defined in the model.

If you want the initial *amount* of a species to be defined inside a NaN-sized compartment, that will, in theory, work OK, but it's still a little dodgy, and there might be bits that go wrong.  I would at least be sure to set 'hasOnlySubstanceUnits' to 'true' for any such species.

**Comment by luciansmith** [comment]
After testing, I can confirm my suspicions:  you must indeed set hasOnlySubstanceUnits to 'true', and you must not request the species concentration.  But if you do that, everything works fine:

```
import tellurium as te

r = te.loada("""
 substanceonly species a in C
 a = 1/C #This sets the initial amount instead of the initial concentration.
 -> a; .1*a
 C = nan
""")

r.timeCourseSelections = ["time", "a"]
r.simulate()
r.plot()

**Comment by matthiaskoenig** [comment]
Sorry for the slow feedback.
- Yes, roadrunner handles species internally as amounts, so if there is a concentration species (`hasOnlySubstanceUnits`) with a NaN compartment size an `initialConcentration` will be converted to NaN.
- My issue is that the `initialConcentration` information is dropped from the model and from the calculation of the initial state (e.g. via `resetAll`).

This issue is part of a larger issue how roadrunner handles the initial state of a model. Basically there are two interpretations what the initial state of a model could be
1. The model after model loading (this is how roadrunner is doing it)
2. The model state consisting of all the `init` conditions. This consists of all `InitialAssignments`, the `initalAmount` and `initalConcentrations` of all species and the values at model loading for everything else such as sizes of compartments. The modeler can then set additional init information with the `init(symbol)` syntax. This would allow to use roadrunner for additional use cases without requiring to create new derived SBML models with the changes (which take long to load). 

I.e. the following would be what would be how things should be handled
```
# loads model, sets all `init` for symbols, applies initalConcentration and initialAmount, and InitialAssignment
rr = RoadRunner(model)  

# set user defined additional `init` conditions (resetAll resets to these inits)
rr.setValue("init(compartment1)") = 10


# Resets time, all floating species, and rates to their CURRENT initial values. 
# Note it is NOT the initial values of when the model was first loaded
rr.resetAll()  # <- This should apply the `initalConcentration` correctly based on all the set init conditions, but the information is lost.
```

So in summary: I understand why this is not working currently and how to workaround this (create new modified SBML and load it), but there could be a better way of handling this, by doing this in the resetAll() function correctly.
I hope this makes sense ;)

**Comment by luciansmith** [comment]
A bunch of updates were made to 'resetAll' since this was first posted--have the issues been resolved?

**Comment by hsauro** [comment]
The resets (other than resetToOrgin) should be taking their init values
from the init symbols, and not from the SBML model.

**Comment by matthiaskoenig** [comment]
I am checking this issue and will provide an update on it.

**Comment by luciansmith** [comment]
@matthiaskoenig : This is very old at this point, but do you have any updates?  Or should we close this issue?

**Comment by matthiaskoenig** [comment]
This is not solved, but can be closed. 
This was an issue which mainly came from model coupling. I solved this by just avoiding the issue. I.e. I set a dummy volume of 1.0 instead of NaN.
Many tools have issues with NaN volumes besides roadrunner, so I just stopped using this ;).
